pkg/apis/migration/migration.go

```diff
@@ -0,0 +1,351 @@
+		crd, err = apiextensionsClientSet.ApiextensionsV1beta1().CustomResourceDefinitions().Get(crdName, metav1.GetOptions{})
+		if err != nil {
+			log.Errorf("failed to get %q CRD with err: %+v", crdName, err)
+			return false, err
+		}
+		for _, cond := range crd.Status.Conditions {
+			switch cond.Type {
+			case apiextensionsv1beta1.Established:
+				if cond.Status == apiextensionsv1beta1.ConditionTrue {
+					return true, err
+				}
+			case apiextensionsv1beta1.NamesAccepted:
+				if cond.Status == apiextensionsv1beta1.ConditionFalse {
+					log.Debugf("Name conflict while waiting for %q CRD creation", cond.Reason)
+				}
+			}
+		}
+		return false, err
+	})
+
+	// If there is an error, delete the object to keep it clean.
+	if err != nil {
+		log.Infof("Cleanup %q CRD because the CRD created was not successfully established. Error: %+v", crdName, err)
+		deleteErr := apiextensionsClientSet.ApiextensionsV1beta1().CustomResourceDefinitions().Delete(crdName, nil)
+		if deleteErr != nil {
+			log.Errorf("failed to delete %q CRD with error: %+v", crdName, deleteErr)
+		}
+	}
+	return err
+}
+
+// GetVolumeInfo returns CnsvSphereVolumeMigrationSpec for given VolumePath from local cache
+// if CnsvSphereVolumeMigrationSpec is not available in the cache, K8s CR look up will be performed, if that too
+// can not retrieve VolumeInfo, ErrVolumeNotRegistered will be returned
+func (volumeMigration *volumeMigration) GetVolumeInfo(ctx context.Context, volumePath string) (migration.CnsvSphereVolumeMigrationSpec, error) {
+	log := logger.GetLogger(ctx)
+	info, found := volumeMigration.volumePathToVolumeInfo.Load(volumePath)
+	if found {
+		log.Infof("CnsvSphereVolumeMigrationSpec found from the cache for for VolumePath: %q", volumePath)
+		return info.(migration.CnsvSphereVolumeMigrationSpec), nil
+	}
+	volumeMigrationCRList := &migration.CnsvSphereVolumeMigrationList{}
+	err := volumeMigration.migrationClient.List(ctx, volumeMigrationCRList)
+	if err != nil {
+		log.Errorf("failed to get volumeMigration CR list")
+		return migration.CnsvSphereVolumeMigrationSpec{}, err
+	}
+	for _, object := range volumeMigrationCRList.Items {
+		if object.Spec.VolumePath == volumePath {
+			log.Infof("found CR for VolumePath: %q", volumePath)
+			volumeMigration.volumePathToVolumeInfo.Store(volumePath, object.Spec)
+			log.Debugf("cached VolumeMigrationSpec for VolumePath: %q", volumePath)
+			return object.Spec, nil
+		}
+	}
+	log.Infof("Could not retrieve CnsvSphereVolumeMigrationSpec from cache and K8s CR for Volume Path: %q. volume may not be registered", volumePath)
+	return migration.CnsvSphereVolumeMigrationSpec{}, ErrVolumeNotRegistered
+}
+
+// SaveVolumeInfo helps create CR for given cnsvSphereVolumeMigration
+// this func also update local cache with supplied cnsvSphereVolumeMigration, after successful creation of CR
+func (volumeMigration *volumeMigration) SaveVolumeInfo(ctx context.Context, cnsvSphereVolumeMigration migration.CnsvSphereVolumeMigration) error {
```

Comment:
Can a pointer be passed into the function instead of the whole object?

Reply:
yes, pointer would be good.